File: Solutions/Python/edit_distance.py
"""
    [MEDIUM]
    72. Edit Distance

    Concepts:
    - string
    - dynamic programming

    Stats:
        Runtime | 63 ms     [Beats 64.56%]
        Memory  | 21.18 MB  [Beats 57.98%]
"""

class Solution:
    def minDistance(self, w1: str, w2: str) -> int:
        """
            Imagine 2D matrix where chars of word1
            are along column and chars of word2 are
            along rows
        """
        m, n = len(w1), len(w2) 
        dp = [[float("inf")]*(n+1) for _ in range(m+1)]

        # initialize empty rows and cols for case
        # when either string is empty
        for j in range(n + 1):
            dp[m][j] = n - j
        for i in range(m + 1):
            dp[i][n] = m - i

        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                if w1[i] == w2[j]:
                    dp[i][j] = dp[i+1][j+1]
                else:
                    dp[i][j] = 1 + min(
                        dp[i+1][j],     # delete
                        dp[i][j+1],     # insert
                        dp[i+1][j+1]    # replace
                    )

        return dp[0][0]

File: Solutions/Cpp/edit_distance.cpp
/*
    [MEDIUM]
    72. Edit Distance

    Concepts:
    - string
    - dynamic programming
    
    Stats:
        Runtime | 4 ms      [Beats 80.30%]
        Memory  | 12.93 MB  [Beats 73.77%]
*/

class Solution {
public:
    int minDistance(string w1, string w2) {
        /*
            Imagine 2D matrix where chars of word1
            are along column and chars of word2 are
            along rows
        */
        const int m = w1.size(), n = w2.size();
        std::vector<std::vector<int>> dp(
            m+1,
            std::vector<int>(
                n+1,
                std::numeric_limits<int>::infinity()
            )
        );

        // initialize empty rows and cols for case
        // when either string is empty
        for (int j = 0; j <= n; j++)
            dp[m][j] = n - j;
        for (int i = 0; i <= m; i++)
            dp[i][n] = m - i;

        for (int i = m - 1; i > -1; i--) {
            for (int j = n - 1; j > -1; j--) {
                if (w1[i] == w2[j])
                    dp[i][j] = dp[i+1][j+1];
                else
                    dp[i][j] = 1 + std::min(
                        std::min(
                            dp[i+1][j],     // delete
                            dp[i][j+1]      // insert
                        ),
                        dp[i+1][j+1]        // replace
                    );
            }
        }

        return dp[0][0];
    }
};